File: fieldmate/src/main/java/com/nguyencong/fieldmate/service/BookingService.java
package com.nguyencong.fieldmate.service;

import java.time.LocalDate;
import java.util.List;

import org.springframework.data.domain.Page;

import com.nguyencong.fieldmate.dto.request.BookingRequest;
import com.nguyencong.fieldmate.dto.response.BookingResponse;
import com.nguyencong.fieldmate.entity.enums.BookingStatus;

public interface BookingService {
    BookingResponse createBooking(BookingRequest request);
    List<BookingResponse> getCurrentCustomerBookings();
    BookingResponse getBookingById(Long id);
    Page<BookingResponse> getBookingsByVenueId(Long venueId, LocalDate date, BookingStatus status, Long bookingId, int page);
    BookingResponse completeBooking(Long id);
    Page<BookingResponse> getAllBookings(String search, BookingStatus status, int page);
}

File: front-end/fieldmate-web/src/services/booking.service.ts
import {
  fieldmateClient,
  fieldmateEndpoints,
} from "@/configs/fieldmate-client";
import type {
  BookingRequest,
  BookingResponse,
  BookingStatus,
} from "@/types/booking";
import type { PageResponse } from "@/types/pagination";

type GetVenueBookingsParams = {
  page?: number;
  date?: string;
  status?: BookingStatus;
  bookingId?: number;
};

type GetAdminBookingsParams = {
  search?: string;
  status?: BookingStatus;
  page?: number;
};

export const bookingService = {
  async create(request: BookingRequest): Promise<BookingResponse> {
    const response = await fieldmateClient.post<BookingResponse>(
      fieldmateEndpoints.secureBookings,
      request,
    );

    return response.data;
  },

  async getMyBookings(): Promise<BookingResponse[]> {
    const response = await fieldmateClient.get<BookingResponse[]>(
      fieldmateEndpoints.myBookings,
    );

    return response.data;
  },

  async getById(bookingId: number): Promise<BookingResponse> {
    const response = await fieldmateClient.get<BookingResponse>(
      fieldmateEndpoints.secureBooking(bookingId),
    );

    return response.data;
  },

  async getByVenueId(
    venueId: number,
    params: GetVenueBookingsParams = {},
  ): Promise<PageResponse<BookingResponse>> {
    const response = await fieldmateClient.get<
      PageResponse<BookingResponse>
    >(fieldmateEndpoints.secureVenueBookings(venueId), {
      params: {
        page: params.page ?? 0,
        date: params.date || undefined,
        status: params.status,
        bookingId: params.bookingId,
      },
    });

    return response.data;
  },

  async complete(bookingId: number): Promise<BookingResponse> {
    const response = await fieldmateClient.patch<BookingResponse>(
      fieldmateEndpoints.completeBooking(bookingId),
    );

    return response.data;
  },

  async getAll(
    params: GetAdminBookingsParams = {},
  ): Promise<PageResponse<BookingResponse>> {
    const response = await fieldmateClient.get<
      PageResponse<BookingResponse>
    >(fieldmateEndpoints.secureBookings, {
      params: {
        search: params.search?.trim() || undefined,
        status: params.status,
        page: params.page ?? 0,
      },
    });

    return response.data;
  },
};
